File: src/compiler/lexer/token/IdentifierToken.java
package compiler.lexer.token;

import compiler.parser.TokenEvaluator;
import compiler.parser.visitors.TokenVisitor;

public class IdentifierToken extends TypedToken<String> {
  private static final IdentifierToken sentinel = new IdentifierToken("Sentinel");

  public IdentifierToken(String s) {
    super(s, 2);
  }

  @Override
  protected String parse(String str) {
    return str;
  }

  public static IdentifierToken getSentinel() {
    return sentinel;
  }

  @Override
  public <T> T accept(TokenVisitor<T> visitor) {
    return visitor.visit(this);
  }

  @Override
  public Object accept(TokenEvaluator visitor) throws Exception {
    return visitor.visit(this);
  }
}

File: src/compiler/lexer/token/IdentifierToken.kt
package compiler.lexer.token

import compiler.parser.TokenEvaluator
import compiler.parser.visitors.TokenVisitor

class IdentifierToken(s: String) : TypedToken<String>(s, 2) {
    override fun parse(str: String): String {
        return str
    }

    override fun <T> accept(visitor: TokenVisitor<T>): T {
        return visitor.visit(this)
    }

    @Throws(Exception::class)
    override fun accept(visitor: TokenEvaluator): Any {
        return visitor.visit(this)
    }

    companion object {
        val sentinel = IdentifierToken("Sentinel")
    }
}
